build_defs/sdk/sdk.build_defs

```diff
@@ -0,0 +1,142 @@
+def proto_build_defs(name:str, srcs:str|list, proto_languages:dict, visibility=["PUBLIC"]):
+    """Exposes `.build_defs` containing `proto_language()` rules
+
+    name (str): The name of this rule
+    src (str|list): The .build_defs files containing the definition
+    proto_languages (dict): A map of proto language types to functions that return the `proto_language()` below.
+                            e.g. `{"grpc_language": ["go_grpc_language"], "proto_language": ["go_proto_language"]}`
+                            where `go_grpc_language` is a function that has a `return proto_language(...)`
+    """
+    labels = []
+    for type, langs in proto_languages.items():
+        labels += [f"{type}:{lang}" for lang in langs]
+    return filegroup(
+       name = name,
+       srcs = srcs,
+       visibility = visibility,
+       labels = labels,
+   )
+
+def proto_language(language:str, build_def:function, additional_provides:list|dict=[]):
+    """Returns the definition of how to build a particular language for proto_library or grpc_library.
+
+    This should be wrapped in a function so that the proto_languages() build definition below can 
+    
+    Args:
+      language (str): Name of the language (as we would name it).
+      build_def (function): The build definition lambda. Will be called with the same arguments proto_library
+                            was called with, except with an additional `tag` parameter which is to be applied to the
+                            final build rule. Should return
+      additional_provides (list|dict): Any additional tags from the rule generated by func that should be provided by
+                                       the final proto_library rule. The _{name}#{language} tag will already be provided.
+                                       Can be a dictionary if the provide key on the `proto_library` should differ from
+                                       the language rule. In which case, the key will be the key on the proto_library,
+                                       and the value will be the tag to provide from build_def.
+    """
+    return {
+        'language': language,
+        'build_def': build_def,
+        'additional_provides': additional_provides,
+    }
+
+def _get_languages_prebuild(label_prefix, includes):
```

Comment:
This outputs a series of echo commands that build up the build definition file that creates the `proto_languages()` function that `proto_library()` etc. can use. 

This is a little gross but could probably be cleaned up with a tool rather than doing the templating here. I don't think it's that bad considering the value it adds.